producer-kafka/src/main/java/io/stargate/producer/kafka/schema/SchemaRegistryProvider.java

```diff
@@ -0,0 +1,235 @@
+      Schema avroFieldSchema = CqlToAvroTypeConverter.toAvroType(columnMetadata.getType());
+      keyBuilder.name(columnMetadata.getName()).type(avroFieldSchema).noDefault();
+    }
+    return keyBuilder.endRecord();
+  }
+
+  Schema constructValueSchema(TableMetadata tableMetadata) {
+    List<Schema> partitionKeys =
+        constructRequiredValueFieldsSchema(tableMetadata.getPartitionKeys());
+    List<Schema> clusteringKeys =
+        constructRequiredValueFieldsSchema(tableMetadata.getClusteringKeys());
+    List<Schema> columns = constructOptionalValueFieldsSchema(tableMetadata.getColumns());
+    Schema fieldsSchema =
+        constructFieldsSchema(partitionKeys, clusteringKeys, columns, tableMetadata);
+
+    String valueRecordName = constructValueRecordName(tableMetadata);
+    return SchemaBuilder.record(valueRecordName)
+        .fields()
+        .requiredString(OPERATION_FIELD_NAME)
+        .name(TIMESTAMP_FIELD_NAME)
+        .type(TIMESTAMP_MILLIS_TYPE)
+        .noDefault()
+        .name(DATA_FIELD_NAME)
+        .type(fieldsSchema)
+        .noDefault()
+        .endRecord();
+  }
+
+  private Schema constructFieldsSchema(
+      List<Schema> partitionKeys,
+      List<Schema> clusteringKeys,
+      List<Schema> columns,
+      TableMetadata tableMetadata) {
+    String dataRecordName = constructDataRecordName(tableMetadata);
+    FieldAssembler<Schema> fields = SchemaBuilder.record(dataRecordName).fields();
+    addToFields(partitionKeys, fields);
+    addToFields(clusteringKeys, fields);
+    addToFields(columns, fields);
+    return fields.endRecord();
+  }
+
+  private void addToFields(List<Schema> fieldsToAdd, FieldAssembler<Schema> fields) {
+    for (Schema schema : fieldsToAdd) {
+      if (!schema.getType().equals(Type.UNION)) {
+        throw new IllegalStateException(
+            String.format("The type for %s should be UNION but is: %s", schema, schema.getType()));
+      }
+      String fieldName = schema.getTypes().get(1).getName();
+      fields.name(fieldName).type(schema).withDefault(null);
+    }
+  }
+
+  private List<Schema> constructRequiredValueFieldsSchema(List<ColumnMetadata> tableMetadata) {
+    List<Schema> fields = new ArrayList<>();
+    for (ColumnMetadata columnMetadata : tableMetadata) {
+      Schema partitionKey =
+          SchemaBuilder.record(columnMetadata.getName())
+              .fields()
+              .name(VALUE_FIELD_NAME)
+              .type(CqlToAvroTypeConverter.toAvroType(columnMetadata.getType()))
+              .noDefault()
+              .endRecord();
+      fields.add(SchemaBuilder.unionOf().nullType().and().type(partitionKey).endUnion());
+    }
+    return fields;
+  }
+
+  private List<Schema> constructOptionalValueFieldsSchema(List<ColumnMetadata> tableMetadata) {
+    List<Schema> partitionKeys = new ArrayList<>();
+    for (ColumnMetadata columnMetadata : tableMetadata) {
+      Schema partitionKey =
```

Comment:
I think the variable names "partitionKeys" don't apply here, probably a leftover.

Reply:
good catch, done